tornado/routing.py

```diff
@@ -0,0 +1,335 @@
+
+                delegate = self.get_target_delegate(
+                    rule.target, request, **target_params)
+
+                if delegate is not None:
+                    return delegate
+
+        return None
+
+    def get_target_delegate(self, target, request, **target_params):
+        if isinstance(target, Router):
+            route = target.find_handler(request, **target_params)
+            if route is not None:
+                return route
+
+        elif callable(target):
+            return _CallableAdapter(
+                partial(target, **target_params), request.connection
+            )
+
+        return None
+
+
+class ReversibleRuleRouter(ReversibleRouter, RuleRouter):
+    def __init__(self, rules=None):
+        self.named_rules = {}
+        super(ReversibleRuleRouter, self).__init__(rules)
+
+    def process_rule(self, rule):
+        rule = super(ReversibleRuleRouter, self).process_rule(rule)
+
+        if rule.name:
+            if rule.name in self.named_rules:
+                app_log.warning(
+                    "Multiple handlers named %s; replacing previous value",
+                    rule.name)
+            self.named_rules[rule.name] = rule
+
+        return rule
+
+    def reverse_url(self, name, *args):
+        if name in self.named_rules:
+            return self.named_rules[name].matcher.reverse(*args)
+
+        for rule in self.rules:
+            if isinstance(rule.target, ReversibleRouter):
+                reversed_url = rule.target.reverse_url(name, *args)
+                if reversed_url is not None:
+                    return reversed_url
+
+        return None
+
+
+class Rule(object):
+    def __init__(self, matcher, target, target_kwargs=None, name=None):
+        self.matcher = matcher
+        self.target = target
+        self.target_kwargs = target_kwargs if target_kwargs else {}
+        self.name = name
+
+    def reverse(self, *args):
+        if isinstance(self.matcher, Matcher):
+            return self.matcher.reverse(*args)
+        return None
+
+    def __repr__(self):
+        return '%s(%r, %s, kwargs=%r, name=%r)' % \
+               (self.__class__.__name__, self.matcher,
+                self.target, self.target_kwargs, self.name)
+
+
+class Matcher(object):
+    """A complex matcher can be represented as an instance of some
+    `Matcher` subclass. It must implement ``__init__`` (for example
+    to initialize a regex pattern) and ``__call__`` (which is called
+    with a `httpserver.HTTPRequest` argument).
+    """
+
+    def __call__(self, request):
+        """Matches an instance against the request.
+
+        :returns a dict of parameters (handler_kwargs, path_args, path_kwargs)
+        to be passed to the target handler"""
+        raise NotImplementedError()
+
+    def reverse(self, *args):
+        """Reconstruct URL from matcher instance"""
+        return None
+
+
+ANY_MATCHES = lambda request: {}
```

Comment:
Also fixed in https://github.com/tornadoweb/tornado/pull/1806/commits/4c41a1c06f26c5913f24e2d56d4639abd9a236d4